hash/hash_comparator.go

```diff
@@ -1,55 +1,160 @@
+var ErrUnknownHashAlgorithm = fmt.Errorf("unknown hash algorithm")
+var ErrUnknownHashFormat = fmt.Errorf("unknown hash format")
+var ErrEmptyHashCompare = fmt.Errorf("empty hash provided")
+var ErrEmptyPasswordCompare = fmt.Errorf("empty password provided")
 
-func Compare(ctx context.Context, password []byte, hash []byte) error {
-	if IsBcryptHash(hash) {
-		return CompareBcrypt(ctx, password, hash)
-	} else if IsArgon2idHash(hash) {
-		return CompareArgon2id(ctx, password, hash)
-	} else {
+var hashSeparator = []byte("$")
+var bcryptLegacyPrefix = regexp.MustCompile(`^2[abzy]?$`)
+
+func Compare(ctx context.Context, cfg *config.Config, password []byte, hash []byte) error {
+	algorithm, realHash, err := ParsePasswordHash(hash)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	switch {
+	case bytes.Equal(algorithm, Argon2AlgorithmId):
+		return CompareArgon2id(ctx, cfg, password, realHash)
+	case bytes.Equal(algorithm, BcryptAlgorithmId):
+		return CompareBcrypt(ctx, cfg, password, realHash)
+	case bytes.Equal(algorithm, BcryptAESAlgorithmId):
+		return CompareBcryptAes(ctx, cfg, password, realHash)
+	default:
 		return ErrUnknownHashAlgorithm
 	}
 }
 
-func CompareBcrypt(_ context.Context, password []byte, hash []byte) error {
+func CompareBcrypt(_ context.Context, _ *config.Config, password []byte, hash []byte) error {
 	if err := validateBcryptPasswordLength(password); err != nil {
-		return err
+		return errors.WithStack(err)
 	}
 
 	err := bcrypt.CompareHashAndPassword(hash, password)
 	if err != nil {
-		return err
+		return errors.WithStack(err)
 	}
 
 	return nil
 }
 
-func CompareArgon2id(_ context.Context, password []byte, hash []byte) error {
+func IsPasswordHash(hash []byte) bool {
+	_, _, err := ParsePasswordHash(hash)
+	return err == nil
+}
+
+func ParsePasswordHash(input []byte) (algorithm, hash []byte, err error) {
+	hashParts := bytes.SplitN(input, hashSeparator, 3)
+	if len(hashParts) != 3 {
+		err = errors.WithStack(ErrUnknownHashFormat)
+		return
+	}
+
+	hash = append(hashSeparator, hashParts[2]...)
+	switch {
+	case bytes.Equal(hashParts[1], Argon2AlgorithmId):
+		algorithm = Argon2AlgorithmId
+		return
+	case bcryptLegacyPrefix.Match(hashParts[1]):
+		hash = input
+		fallthrough
+	case bytes.Equal(hashParts[1], BcryptAlgorithmId):
+		algorithm = BcryptAlgorithmId
+		return
+	case bytes.Equal(hashParts[1], BcryptAESAlgorithmId):
+		algorithm = BcryptAESAlgorithmId
+		return
+	default:
+		err = ErrUnknownHashAlgorithm
+		return
+	}
+}
+
+func aes256Decrypt(data, key []byte) ([]byte, error) {
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+
+	gcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
+
+	decoded := make([]byte, hex.DecodedLen(len(data)))
+	_, err = hex.Decode(decoded, data)
+	if err != nil {
+		return nil, err
+	}
+
+	nonce := decoded[:gcm.NonceSize()]
+	ciphertext := decoded[gcm.NonceSize():]
+
+	return gcm.Open(nil, nonce, ciphertext, nil)
```

Comment:
I would feel more comfortable if we used https://github.com/gtank/cryptopasta/blob/master/encrypt.go

Reply:
Done in c5b5e31e